src/channel.rs

```diff
@@ -44,35 +63,147 @@ fn with_fuse_args<T, F: FnOnce(&fuse_args) -> T>(options: &[&OsStr], f: F) -> T
+        let fuse_device_name = "/dev/fuse";
+
+        let fd = match std::fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open(fuse_device_name)
+        {
+            Ok(file) => file.into_raw_fd(),
+            Err(error) => {
+                if error.kind() == io::ErrorKind::NotFound {
+                    error!("{} not found. Try 'modprobe fuse'", fuse_device_name);
+                }
+                return Err(error);
+            }
+        };
+
+        let code = unsafe { libc::fcntl(fd, libc::F_SETFD, libc::FD_CLOEXEC) };
+        if code == -1 {
+            error!("fcntl command failed with {}", code);
+            return Err(io::Error::last_os_error());
+        }
+
+        let code = unsafe { libc::ioctl(fd, FUSE_DEV_IOC_CLONE, &root_fd.0) };
+        if code == -1 {
+            error!("Clone command failed with {}", code);
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(SubChannel::new(FileDescriptorRawHandle(fd), false)?)
+    }
+
+    // mac fuse seems to just re-use the root fd relying onthe atomic semantics setup in the driver
+    // This will have lowerthroughput than the linux approach.
+    #[cfg(target_os = "macos")]
+    fn create_worker(root_fd: &FileDescriptorRawHandle) -> io::Result<SubChannel> {
+        SubChannel::new(*root_fd, true)
+    }
+    ///
+    /// Create worker fd's takes the root/session file descriptor and makes several clones
+    /// This allows file systems to work concurrently over several buffers/descriptors for concurrent operation.
+    /// More detailed description of the protocol is at:
+    /// https://john-millikin.com/the-fuse-protocol#multi-threading
+    ///
+    fn create_sub_channels(
+        root_fd: &FileDescriptorRawHandle,
+        worker_channels: usize,
+    ) -> io::Result<Vec<SubChannel>> {
+        let mut res = Vec::default();
+
+        res.push(SubChannel::new(*root_fd, false)?);
+
+        for _ in 0..worker_channels {
+            res.push(Channel::create_worker(root_fd)?);
+        }
+
+        Ok(res)
+    }
+
+    fn internal_new_on_err_cleanup(
+        mountpoint: PathBuf,
+        worker_channel_count: usize,
+        fd: FileDescriptorRawHandle,
+        fuse_session: *mut c_void,
+    ) -> io::Result<Channel> {
+        Ok(Channel {
+            mountpoint,
+            sub_channels: Arc::new(Channel::create_sub_channels(&fd, worker_channel_count)?),
+            session_fd: fd,
+            fuse_session,
+        })
+    }
+    fn new_from_session_and_fd(
+        mountpoint: &Path,
+        worker_channel_count: usize,
+        fd: FileDescriptorRawHandle,
+        fuse_session: *mut c_void,
+    ) -> io::Result<Channel> {
+        match Channel::internal_new_on_err_cleanup(
+            mountpoint.to_owned(),
+            worker_channel_count,
+            fd,
+            fuse_session,
+        ) {
+            Ok(r) => Ok(r),
+            Err(err) => {
```

Comment:
After more iterations, i'm not sure the inner function here is adding much value at this point -- but it can fail via 

```
Channel::create_sub_channels(&fd, worker_channel_count)?
```

The design/approach of the 2 layers was to just allow the inner function to sprinkle `?` around everywhere and then the outer function would clean up. Now, though  only one spot is fallible, so i'll collapse these and handle directly failure to create the sub channels.